src/Features/CSharp/Portable/CodeRefactorings/PushMemberUp/AbstractMemberPusher.cs

```diff
@@ -0,0 +1,149 @@
+                case FieldDeclarationSyntax fieldDeclarationSyntax:
+                    return CreateFieldPushUpSyntax(fieldDeclarationSyntax, variableDeclaratorSyntax);
+                default:
+                    throw new ArgumentException($"{nameof(memberDeclarationSyntax)}'s type is invalid");
+            }
+        }
+
+        internal virtual CodeAction ComputeRefactoring()
+        {
+            if (UserSelectedNode is VariableDeclaratorSyntax selectVariableDeclaratorNode &&
+                AreModifiersValid(selectVariableDeclaratorNode))
+            {
+                // Event and Field are different, it may contains several declarations in one line
+                var selectFieldOrEventNode = FindFieldAndEventDeclaration(selectVariableDeclaratorNode);
+                return CreateAction(selectFieldOrEventNode, ContextDocument, selectVariableDeclaratorNode);
+            }
+            else if (UserSelectedNode is MethodDeclarationSyntax selectMethodNode &&
+                    AreModifiersValid(selectMethodNode) &&
+                    IsOrdinaryMethod(selectMethodNode))
+            {
+                return CreateAction(selectMethodNode, ContextDocument);
+            }
+            else if (UserSelectedNode is BasePropertyDeclarationSyntax selectPropertyOrIndexerNode &&
+                    AreModifiersValid(selectPropertyOrIndexerNode))
+            {
+                return CreateAction(selectPropertyOrIndexerNode, ContextDocument);
+            }
+            else
+            {
+                return default;
+            }
+        }
+
+        internal virtual CodeAction CreateAction(
+            MemberDeclarationSyntax selectedMemberNode,
+            Document contextDocument,
+            VariableDeclaratorSyntax selectVariableDeclaratorNode = null)
+        {
+            var nodeToPushUp = GeneratePushUpMemberAndFormatting(selectedMemberNode, selectVariableDeclaratorNode);
+
+            if (TargetSyntaxNode.SyntaxTree == selectedMemberNode.SyntaxTree)
+            {
+                return CreateDocumentChangeAction(nodeToPushUp, contextDocument);
+            }
+            else
+            {
+                return CreateSolutionChangeAction(nodeToPushUp, contextDocument);
+            }
+        }
+
+        protected virtual EventFieldDeclarationSyntax CreateEventPushUpSyntax(EventFieldDeclarationSyntax eventFieldDeclarationSyntax, VariableDeclaratorSyntax variableDeclaratorNode)
+        {
+            var identifierSyntax = variableDeclaratorNode.Parent.DescendantNodes().OfType<IdentifierNameSyntax>().First();
+            var declarationList = new SeparatedSyntaxList<VariableDeclaratorSyntax>();
+            declarationList = declarationList.Add(variableDeclaratorNode);
+            var variableDeclaration = SyntaxFactory.VariableDeclaration(identifierSyntax, declarationList);
+            return SyntaxFactory.EventFieldDeclaration(variableDeclaration);
```

Comment:
all this code seems unnecessary.   We already have ICodeGenerationService which already has methods like AddEventAsync.  This should all go away with that code.  (this also applies to all lower code).